[2단계 - 경로 조회 기능] 글렌(전석진) 미션 제출합니다.

seokjin8678 opened the pull request:
안녕하세요 토니!
2단계 기능 요구 사항 완수하고 미션 제출합니다!

2단계 요구 사항에서 최단 경로 탐색 기능은 외부 라이브러리를 사용하는 부분이라 크게 어려운 것은 없었네요!
다만 이번 미션에서 조금 어려웠던 점은 요금 정책이 아닐까 싶어요!

`PricePolicy` 라는 인터페이스를 만들어, `BasicPricePolicy`를 구현체로 사용했습니다!
추후 새로운 금액에 대한 변경 사항이 있으면 쉽게 바꿀 수 있도록 초점을 맞췄네요!

PathService가 PricePolicy를 의존하고 있는데, PriceService를 만들어 PathService가 PriceService를 의존하게 만들어야 할 지 모르겠네요!

리뷰 기다리겠습니다!
감사합니다!

seokjin8678 committed:
build: JGraphT 라이브러리 추가

seokjin8678 committed:
docs: 기능 요구 사항 추가

seokjin8678 committed:
build: task copyDocument -> copyApiDocument로 변경

seokjin8678 committed:
refactor: Domain 패키지 분류

seokjin8678 committed:
feat: PathController 슬라이스 테스트 추가, RestDocs 적용

seokjin8678 committed:
build: JGraphT 라이브러리 수정

- 잘못된 라이브러리 추가 수정

seokjin8678 committed:
feat: 최단 경로를 구하는 PathService 추가

seokjin8678 committed:
fix: 구간을 추가할 때 상행역과 하행역이 같으면 예외가 발생하도록 수정

seokjin8678 committed:
fix: 경로를 조회할 때 HTTP GET -> POST로 메서드 수정

seokjin8678 committed:
docs: 기능 요구 사항 체크

seokjin8678 committed:
refactor: PathResponse totalPrice -> price로 변경

seokjin8678 committed:
feat: 요금 계산 클래스 추가

seokjin8678 committed:
refactor: PathRequest 역으로 끝나는지 검증 추가, Service 로직 리팩터링

seokjin8678 committed:
test: PathController 통합 테스트 추가

seokjin8678 committed:
refactor: 패키지 구조 세분화

seokjin8678 committed:
build: MariaDB JDBC 커넥터 추가

seokjin8678 committed:
feat: 프로덕션, 테스트 DB 환경 분리

seokjin8678 committed:
fix: EndsWith 검증에 null 검사 추가

seokjin8678 committed:
refactor: 역 이름, 노선 이름 최소 길이 제한 조건 추가

seokjin8678 committed:
docs: 기능 구현 목록 체크

seokjin8678 committed:
test: 테스트 이름에 요금 추가

seokjin8678 committed:
refactor: 요금 계산 로직 리팩터링

toneyparky commented on README.md:
````diff
@@ -1,167 +1,6 @@
-    }
-  ]
-  ```
+### [API 명세](http://localhost:8080/docs/html5/index.html)
````
굳굳~! 이제 프로젝트에 참여하는 개발자가 쉽게 API 명세를 확인할 수 있겠네요! 프로젝트가 제공하는 API를 확인하기 쉬워서 협업하기도 좋겠어요 😬
지금은 아니지만 나중에 프로젝트를 배포하게 되면 배포한 주소에 대해서 API가 생성되겠죠!

toneyparky commented on src/main/java/subway/application/path/BasicPricePolicy.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.Map;
+import java.util.function.IntUnaryOperator;
+
+public class BasicPricePolicy implements PricePolicy {
```
기본 가격 정책이란건 뭘까요? 해당 정책 내부에서는 어떤 기준으로 정책을 정하는지가 이름으로 드러나면 더 좋지 않을까해요. 다른 정책이 추가될 경우에도 이 이름이 유지될거라 생각하지 않기 때문이에요. 이건 3단계를 진행하게 되면서 고려해봐도 좋습니다~!

seokjin8678 replied:
우선 내부에선 거리 별 차등으로 가격을 확인하네요!
추후 더 나은 이름이 있는지 고려해보겠습니다!

toneyparky replied:
차등 거리 가격 정책으로 이름이 변경되었네요~! 현재는 가격 정책이 거리 뿐이지만 3단계에서는 더 늘어날테니 그에 맞게 리팩터링해보기로하죠..!

toneyparky commented on src/main/java/subway/application/path/BasicPricePolicy.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.function.IntUnaryOperator;
+
+public class BasicPricePolicy implements PricePolicy {
+    private static final int LEVEL1_PRICE = 1250;
+    private static final int LEVEL2_MAX_PRICE = 2050;
+    private static final int EXTRA_PRICE_RATE = 100;
+    private static final int LEVEL1_THRESHOLD = 11;
+    private static final int LEVEL2_THRESHOLD = 51;
+    private static final int LEVEL2_PER_DISTANCE = 5;
+    private static final int LEVEL3_PER_DISTANCE = 8;
```
`PriceLevel`이라는 객체는 어떤 책임을 가지나요? 

이 값들의 일부는 `PriceLevel`의 책임일수도 있겠다 싶어요.

seokjin8678 replied:
PriceLevel 객체의 책임을 더 명확히 변경해보겠습니다!

toneyparky replied:
제가 의도한 바대로 수정해주셨다고 생각되어요 👍

`DistanceLevel`를 `DifferentialDistancePricePolicy` 내에서만 사용한다면 정책 객체의 내부에 private 클래스로 두는 것도 방법이라 생각해요~! 연관된 두 코드를 같은 파일에 두는거죠!

나중에 코틀린을 접하게 되신다면 하나의 파일에 여러 클래스를 선언하여 응집도를 높이는 패턴을 많이 보실거에요. 지금은 참고만 해주십셥~

toneyparky commented on src/test/java/subway/application/path/BasicPricePolicyTest.java:
```diff
@@ -0,0 +1,81 @@
+     */
+    @ParameterizedTest
+    @CsvSource(value = {"51:2150", "58:2150", "59:2250", "66:2250", "74:2350", "98:2650"}, delimiter = ':')
+    @DisplayName("거리가 50km를 초과하면 8km 마다 1000원이 추가되어야 한다.")
```
```suggestion
    @DisplayName("거리가 50km를 초과하면 8km 마다 100원이 추가되어야 한다.")
```

올바르지 않은 설명을 수정했어요.

seokjin8678 replied:
꼼꼼하게 봐주셨군요..!
👍👍👍👍👍

toneyparky commented on src/test/java/subway/application/path/PathServiceTest.java:
```diff
@@ -0,0 +1,141 @@
+
+    PricePolicy pricePolicy = mock(PricePolicy.class);
+
+    /*
+     *   [A역]-7-[B역]-8-[C역]
+     *     |             |
+     *     5             1
+     *     |             |
+     *   [D역]-2-[F역]-3-[G역]
+     *     |             |
+     *     8             1
+     *     |             |
+     *   [E역]-4-[H역]-2-[I역]
+     *
+     *   [X역]-1-[Z역]
```
이렇게 다이어그램을 그려주시니 이해하기가 훨씬 편하네요..! 굳굳

seokjin8678 replied:
종이에 그리려고 했지만, 테스트 코드에 눈으로 보여야 이해하기 쉬울 것 같아서 조금 노력해봤네요 😂

toneyparky replied:
핵심 로직이고 테스트 코드를 파악하기 쉬워야하기에 이런 부분은 좋은 센스입니다~!

toneyparky commented on src/main/java/subway/config/PriceConfiguration.java:
```diff
@@ -0,0 +1,15 @@
+import subway.application.path.BasicPricePolicy;
+import subway.application.path.PricePolicy;
+
+@Configuration
+public class PriceConfiguration {
+
+    @Bean
+    public PricePolicy basicPricePolicy() {
+        return new BasicPricePolicy();
+    }
+}
```
`@Configuration`는 어떤 기능을 하는 어노테이션이고 언제 사용해야해요? `BasicPricePolicy`를 그냥 `@Component`로 등록하면 안되는걸까요?

seokjin8678 replied:
스프링에서 빈을 등록하는 방법은 여러가지가 있는데, 그 중 자주 사용하는 방법은 2가지 입니다!
1. 자동 빈 등록(컴포넌트 스캔)
2. 수동 빈 등록(`@Configuraiton`)

보통의 경우 컴포넌트 스캔을 통해 자동 빈 등록을 사용하지만, 관리할 빈들이 많아지고 클래스의 갯수가 늘어난다면 컴포넌트가 어디 클래스에 붙어있는지 명확하게 알기가 어려워진다고 생각합니다!

특히, `PricePolicy`의 경우 인터페이스이고 해당 인터페이스를 구현한 다른 요금 정책이 계속 생겨난다면 클래스마다 컴포넌트 에너테이션을 붙이는 것 보다, 수동 빈 등록을 통해 한 클래스 파일에서 관리하는게 추후 가독성이나 유지보수 측면에서 더 이점이 생겨난다고 판단했습니다!

toneyparky replied:
오케이~! 정책이 늘어날 경우를 대비하셨다고 이해했습니다. 3단계에서 효용을 지켜보죠!

toneyparky commented on src/main/java/subway/ui/PathController.java:
```diff
@@ -0,0 +1,31 @@
+import subway.dto.response.Response;
+
+@RestController
+@RequestMapping("/path")
```
다른 API 경로와 달리 path는 단수형으로 지정해주신 이유가 궁금해요~!

seokjin8678 replied:
경로는 아무래도 하는 일이 ~역에서 ~역으로 가는 길찾기 라는 행위를 하는데, 복수형으로 한다면 저장된 경로들을 보여주고, 수정하는 등 CRUD 같은 기능을 제공하나? 라고 생각할 우려가 있어서 입니다!
RESTful API에선 자원들을 대부분 복수형으로 제공하던데, 사실 지금 github의 URI만 보더라도, pull이 단수형인걸 보면 그닥 중요한 요소는 아닌 것 같네요 😂

toneyparky replied:
크... 맞습니다 모든 API가 REST 일 수는 없어요. path가 자원이 아니라고 바라볼 수도 있구요. 핵심은 이해하기 쉽고 명확하게 표현하는거라 생각됩니다. 팀 프로젝트를 진행하며 혹은 다음 미션에서 클라이언트와 열심히 소통하고 협업하시길 바라요..! 지금 같은 경우는 `/shortest-path` 로 표현했다면 더 의도를 잘 드러냈겠다도 싶네요!

toneyparky commented on src/main/java/subway/ui/PathController.java:
```diff
@@ -0,0 +1,31 @@
+        this.pathService = pathService;
+    }
+
+    @PostMapping
+    public ResponseEntity<Response> findPath(@RequestBody PathRequest pathRequest) {
+        PathResponse pathResponse = pathService.findPath(pathRequest);
+        return Response.ok()
+                .message(pathRequest.getOriginStationName() + "에서 " + pathRequest.getDestinationStationName()
+                        + "까지의 경로가 조회되었습니다.")
+                .result(pathResponse)
+                .build();
+    }
```
조회를 위한 API의 메서드가 Post인 이유가 있을까요~?

seokjin8678 replied:
안 그래도 931048db749bfa59d80296c0748a2c42c265dac6 해당 커밋에서 GET으로 했다가, POST로 바꿨는데요!
경로 조회의 경우 `@RequestBody`로 출발역과 도착역을 받는데, GET의 경우 BODY로 데이터를 전송하는 것은 좋지 않다고 들어, POST로 수정하였습니다!
다만, GET을 사용하고 `@RequestParam`으로 값을 받아오면 어땠을지 싶네요!
이렇게 된다면, 해당 경로의 URL을 누군가에 공유할 수도 있겠군요!

toneyparky replied:
저도 쿼리 파라미터가 더 낫다 싶네요..! 공유하는 기능도 유용할 것 같구요! 
상용 웹 사이트를 보면 생각보다 많은 곳에서 쿼리 파라미터를 사용하여 URL을 구성한답니다~!

toneyparky commented on src/main/java/subway/ui/SectionController.java:
```diff
@@ -37,12 +39,19 @@ public ResponseEntity<Response> findAllSectionsByLineId(@PathVariable Long lindI
     @PostMapping("/{lineId}")
     public ResponseEntity<Response> createSection(@PathVariable Long lineId,
                                                   @RequestBody @Valid SectionCreateRequest sectionCreateRequest) {
+        validateSameStation(sectionCreateRequest);
```
오..! 놓친 부분이 있었군요. 잘 챙겨주셨네요 굳굳!

다만, 해당 유효성 검사의 내용이 이 API를 사용하는 클라이언트에만 적용되어야할까요? 혹은 모든 도메인 로직에 다 적용되어야할까요? 만일 후자의 케이스지만 지금처럼 UI 레이어에서 처리하는 상황이라면 새로운 클라이언트에서 구간을 추가하고 싶을 경우 새로운 UI 단의 구현 클래스에서도 이 부분을 또 구현해줘야겠군요?

seokjin8678 replied:
서비스 레이어로 검증 로직을 옮겨봐야겠군요..!
근데 지금 든 생각은 `Section` 도메인에 검증 로직을 추가하여 해결할 수 있을것 같은데요!
제가 구현한 코드에서 도메인은 검증 시 `IllegalArgumentException` 예외를 발생시킵니다!
하지만 `ControllerAdvice`에서 `RuntimeException`의 경우 500번 예외를 발생시키기 때문에, 도메인에서 발생한 예외는 커스텀 예외로 바꾸는게 나은것 같네요!

toneyparky replied:
오호 저는 서비스 로직을 생각했는데 보다 적절한 도메인에 넣어주셨네요 💯

toneyparky commented on src/main/java/subway/application/path/BasicPricePolicy.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public int calculate(int distance) {
+        return calculatePrice(distance);
```
로직이 복잡하거나 여러 단계로 구성되어 있지 않음에도 메서드로 추출해주신 이유가 궁금해요. 글렌이 메서드를 분리하는 기준을 공유해주실 수 있을까요?

seokjin8678 replied:
원래 `PriceLevel`을 if문을 사용하여 분기를 하려고 했는데, 중간에 EnumMap을 사용하여 분기를 처리하는 생각이 떠올라서 적용하는 과정에서 해당 메서드를 분리해놓은 것을 잊었군요 😂

toneyparky replied:
수정 확인했습니다~!

toneyparky commented on src/main/java/subway/application/line/LineService.java:
```diff
@@ -1,4 +1,4 @@
-package subway.application;
+package subway.application.line;
```
패키지를 도메인에 따라 상세하게 분리해주셨네요!
이해하기 더 쉬워졌어요 👍

seokjin8678 replied:
진작 이렇게 분리를 해놓을걸 후회가 드네요 😂

toneyparky replied:
호호 이번 미션에서 패키징에 대해 고민하신 부분에서 조금이나마 기준을 세우셨길 바라요..!

toneyparky reviewed (changes requested):
안녕하세요 글렌~!
리뷰가 조금 늦었지만 어제 이야기 나눈대로 오늘 내에 완료했어요 😊

지난 단계에서 남은 리뷰까지 반영해주시며 요구사항도 잘 구현해주셨네요. 
궁금한 점을 리뷰로 남겼으니 확인 및 답변 부탁드립니다..!

toneyparky commented on src/main/java/subway/application/path/PathService.java:
```diff
@@ -0,0 +1,80 @@
+public class PathService {
+    private final StationDao stationDao;
+    private final SectionDao sectionDao;
+    private final PricePolicy pricePolicy;
```
> PathService가 PricePolicy를 의존하고 있는데, PriceService를 만들어 PathService가 PriceService를 의존하게 만들어야 할 지 모르겠네요

질문 주신 내용을 놓쳤군요!

글렌은 `PriceService`를 만들면 어떤 내용이 들어갈 것 같아요? 즉 `PricePolicy`와 `PriceService`는 어떤 차이가 있나요?

seokjin8678 replied:
해당 내용으로 질문드린건 3단계 요구사항을 미리 보고 어떻게 구조를 잡아야하나 에서 나온 질문입니다!
지금 같은 경우, 거리 별 요금을 책정하지만 3단계에서는 노선 별로 요금을 책정하는 정책이 생기기 때문에, PricePolicy의 인자가 서로 다르게 될 것 같아서 입니다!
DistancePricePolicy(int distance), LinePricePolicy(Line line) 인자가 다른 두 개의 서로 다른 요금 정책이 생길 것 같은데...

이 경우, 하나의 PricePolicy로 거리와 노선 정책을 추상화 할 수 없을거라 판단되어, 모든 요금을 관리하는 `PriceService`를 떠올렸습니다!

toneyparky replied:
아하 2단계의 요구사항만으로는 `PriceService` 책임이 모호하여 질문드렸어요. 고민하신 방법으로 진행해보시죠~!

seokjin8678 committed:
build: Spring REST Docs 누락 설정 추가

- plugin 버전 업그레이드

seokjin8678 committed:
fix: 테스트 이름 오타 수정

seokjin8678 committed:
refactor: 경로 조회 시 HTTP 메소드 POST -> GET으로 변경

- RequestBody로 받는게 아닌, RequestParam으로 출발역, 도착역 정보 수신

seokjin8678 committed:
refactor: BasicPricePolicy 메서드 리팩터링

- 분리된 메서드 다시 합침

seokjin8678 committed:
feat: Section에 같은 역 검증 로직 추가 및 검증 로직 이동

seokjin8678 committed:
fix: 잘못된 거리 측정 수정, PriceLevel -> DistanceLevel으로 이름 변경

seokjin8678 committed:
refactor: 금액 계산 로직 리팩터링

seokjin8678 committed:
chore: 사용하지 않는 Request 클래스 삭제

seokjin8678 committed:
refactor: 가격 정책 이름 변경

- PricePolicy -> DistancePricePolicy
- BasicPricePolicy -> DifferentialDistancePricePolicy

seokjin8678 commented:
안녕하세요 토니!
2단계 피드백 반영하고 다시 제출합니다!

경로를 조회할 때, POST 요청을 GET 요청으로 변경했고(쿼리 파라미터 사용),
가격 정책에 대한 부분을 `거리 가격 정책` 같이 더 명확하도록 변경했습니다!

리뷰 기다리겠습니다!
감사합니다!

seokjin8678 committed:
docs: API 문서 URL 수정

toneyparky reviewed (approved):
안녕하세요 글렌~!

리뷰는 군더더기 없이 잘 반영해주셨습니다. 잘 이해하셨네요..!
DM으로 이야기 나눴듯 3단계로 가보시죠!
하루 즐겁게 마무리하세요 😉

toneyparky reviewed (approved):
안녕하세요 글렌~!

리뷰는 군더더기 없이 잘 반영해주셨습니다. 잘 이해하셨네요..!
DM으로 이야기 나눴듯 3단계로 가보시죠!
하루 즐겁게 마무리하세요 😉
